beater/beater.go

```diff
@@ -203,6 +203,14 @@ func (bt *beater) run(ctx context.Context, cancelContext context.CancelFunc, b *
 				ctx, reload.ReloadableFunc(reloader.reloadOutput),
 			)
 		})
+
+		// Start the manager after all the hooks are initialized
+		// and defined this ensure reloading consistency..
+		if err := b.Manager.Start(); err != nil {
+			return err
+		}
+		defer b.Manager.Stop()
```

Comment:
How[ this worked before](https://github.com/elastic/beats/blob/46be42e7d5a7b87a7511b715f8ce9e1a969f2926/libbeat/cmd/instance/beat.go#L506-L510)? The `SetStopCallBack` concept was existing before, it was just part of the `Manager. Start` and the `beater.Stop` was passed to it, I just separated the Start and Set up of that value.

The logic is the following, if the manager receives a shutdown task from elastic agent, the manager will trigger the callback which is `Beater.Stop`.

Maybe  I need more guidance in your code but from what I understand, this method will `wait` until the beats is stopped when the beat is stopped the manager will be stopped. 

Am I confusing the boot system of apm server?

Reply:
IIANM, SetStopCallBack should never be called by beat-specific code, only by the shared libbeat code during initialisation. It would probably be easier to understand if SetStopCallback was not part of the Manager interface. Instead, pass the stop callback into the manager's constructor?